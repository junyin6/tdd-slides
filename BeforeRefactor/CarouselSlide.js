import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Figure = styled.figure`
  margin: 0;
  padding: 0.5rem;
`;
const Img = styled.img`
  object-fit: cover;
  width: 100%;
  height: ${(props) =>
    typeof props.imgHeight === 'number'
      ? `${props.imgHeight}px`
      : props.imgHeight};
`;
const CarouselSlide = ({
  imgUrl,
  description,
  attribution,
  imgHeight,
  Img,
  ...rest
}) => {
  return (
    <Figure {...rest}>
      <Img src={imgUrl} imgHeight={imgHeight} />
      <figcaption>
        <strong>{description}</strong> {attribution}
      </figcaption>
    </Figure>
  );
};

CarouselSlide.propTypes = {
  Img: PropTypes.elementType,
  imgUrl: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  attribution: PropTypes.node,
  imgHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

CarouselSlide.defaultProps = {
  imgHeight: 600,
  Img: Img,
};

export default CarouselSlide;
