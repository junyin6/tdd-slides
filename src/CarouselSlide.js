import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Img = styled.img`
  object-fit: cover;
  width: 100%;
  height: ${(props) =>
    typeof props.imgHeight === 'number'
      ? `${props.imgHeight}px`
      : props.imgHeight};
`;
const CarouselSlide = ({
  imgUrl,
  description,
  attribution,
  imgHeight,
  ...rest
}) => {
  return (
    <figure {...rest}>
      <Img src={imgUrl} imgHeight={imgHeight} />
      <figcaption>
        <strong>{description}</strong> {attribution}
      </figcaption>
    </figure>
  );
};

CarouselSlide.propTypes = {
  imgUrl: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  attribution: PropTypes.node,
  imgHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

CarouselSlide.defaultProps = {
  imgHeight: 600,
};

export default CarouselSlide;
