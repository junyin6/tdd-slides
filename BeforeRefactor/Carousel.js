import React from 'react';
import CarouselButton from './CarouselButton';
import CarouselSlide from './CarouselSlide';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const CarouselWrapper = styled.div`
  position: relative;
  background-color: black;
  color: #f8f9fa;
`;

const BackButton = styled(CarouselButton)`
  position: absolute;
  margin: 0.5rem;
  padding: 1rem;
  height: 600px;
  left: 0;
  top: 0;
  border: none;
  background: rgba(0, 0, 0, 0.3);
  font-size: 2rem;
  color: #f8f9fa;
  /* top: calc(600px / 2); */
  /* transform: translateY(-50%); */
`;
const ForwardButton = styled(CarouselButton)`
  position: absolute;
  margin: 0.5rem;
  padding: 1rem;
  height: 600px;
  right: 0;
  top: 0;
  border: none;
  background: rgba(0, 0, 0, 0.3);
  font-size: 2rem;
  color: #f8f9fa;
`;

class Carousel extends React.PureComponent {
  static propTypes = {
    slides: PropTypes.arrayOf(PropTypes.shape(CarouselSlide.propTypes))
      .isRequired,
    defaultImgHeight: CarouselSlide.propTypes.imgHeight,
    defaultImg: CarouselSlide.propTypes.Img,
    defaultCarouselWrapper: PropTypes.elementType,
    defaultForwardButton: PropTypes.elementType,
    defaultBackButton: PropTypes.elementType,
  };

  static defaultProps = {
    defaultImgHeight: CarouselSlide.defaultProps.imgHeight,
    defaultImg: CarouselSlide.defaultProps.Img,
    defaultCarouselWrapper: CarouselWrapper,
    defaultForwardButton: ForwardButton,
    defaultBackButton: BackButton,
  };

  state = {
    slideIndex: 0,
  };

  handlePrevClick = () => {
    const { slides } = this.props;
    this.setState(({ slideIndex }) => ({
      slideIndex: (slides.length + slideIndex - 1) % this.props.slides.length,
    }));
  };
  handleNextClick = () => {
    this.setState(({ slideIndex }) => ({
      slideIndex: (slideIndex + 1) % this.props.slides.length,
    }));
  };

  render() {
    const { slides, defaultImgHeight, defaultImg, ...rest } = this.props;
    return (
      <CarouselWrapper {...rest}>
        <CarouselSlide
          Img={defaultImg}
          imgHeight={defaultImgHeight}
          {...slides[this.state.slideIndex]}
        />
        <BackButton data-action="Prev" onClick={this.handlePrevClick}>
          &lt;
        </BackButton>
        <ForwardButton data-action="Next" onClick={this.handleNextClick}>
          &gt;
        </ForwardButton>
      </CarouselWrapper>
    );
  }
}

export default Carousel;
