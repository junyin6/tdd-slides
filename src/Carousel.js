import React from 'react';
import CarouselButton from './CarouselButton';
import CarouselSlide from './CarouselSlide';
import PropTypes from 'prop-types';
class Carousel extends React.PureComponent {
  static propTypes = {
    slides: PropTypes.arrayOf(PropTypes.shape(CarouselSlide.propTypes))
      .isRequired,
  };
  state = {
    slideIndex: 0,
  };

  handlePrevClick = () => {
    const { slides } = this.props;
    this.setState(({ slideIndex }) => ({
      slideIndex: (slides.length + slideIndex - 1) % this.props.slides.length,
    }));
  };
  handleNextClick = () => {
    this.setState(({ slideIndex }) => ({
      slideIndex: (slideIndex + 1) % this.props.slides.length,
    }));
  };
  render() {
    const { slides, ...rest } = this.props;
    return (
      <div {...rest}>
        <CarouselSlide {...slides[this.state.slideIndex]} />
        <CarouselButton data-action="Prev" onClick={this.handlePrevClick}>
          Prev
        </CarouselButton>
        <CarouselButton data-action="Next" onClick={this.handleNextClick}>
          Next
        </CarouselButton>
      </div>
    );
  }
}

export default Carousel;
