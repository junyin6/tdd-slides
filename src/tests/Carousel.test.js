import { shallow } from 'enzyme';
import React from 'react';
import Carousel from '../Carousel';
import CarouselButton from '../CarouselButton';
import CarouselSlide from '../CarouselSlide';

describe('Carousel', () => {
  let wrapper;
  const slides = [
    {
      imgUrl: 'https://example.com/slide1.png',
      description: 'Slide 1',
      attribution: 'Uno Pizzeria',
    },
    {
      imgUrl: 'https://example.com/slide2.png',
      description: 'Slide 2',
      attribution: 'Dos Equis',
    },
    {
      imgUrl: 'https://example.com/slide3.png',
      description: 'Slide 3',
      attribution: 'Three Amigos',
    },
  ];

  beforeEach(() => {
    wrapper = shallow(<Carousel slides={slides} />);
  });

  it('renders a <div>', () => {
    expect(wrapper.type()).toBe('div');
  });

  it('has an initial "slideIndex" of 0', () => {
    expect(wrapper.state('slideIndex')).toBe(0);
  });

  it('renders a CarouselButton labeled "Prev"', () => {
    expect(wrapper.find(CarouselButton).at(0).prop('children')).toBe('Prev');
  });

  it('renders a CarouselButton labeled "Next"', () => {
    expect(wrapper.find(CarouselButton).at(1).prop('children')).toBe('Next');
  });

  describe('with the middle slide selected', () => {
    beforeEach(() => {
      wrapper.setState({ slideIndex: 1 });
    });
    it('increments `slideIndex` when Next is clicked ', () => {
      wrapper.find('[data-action="Next"]').simulate('click');
      expect(wrapper.state('slideIndex')).toBe(2);
    });

    it('decrements `slideIndex` when Prev is clicked', () => {
      wrapper.find('[data-action="Prev"]').simulate('click');
      expect(wrapper.state('slideIndex')).toBe(0);
    });
  });

  describe('with the first slide selected', () => {
    it('wraps to the last slide, when "prev" is clicked', () => {
      wrapper.setState({ slideIndex: 0 });
      wrapper.find('[data-action="Prev"]').simulate('click');
      expect(wrapper.state('slideIndex')).toBe(slides.length - 1);
    });
  });

  describe('with the last slide selected', () => {
    it('wraps to the first slide, when Next is clicked', () => {
      wrapper.setState({ slideIndex: 2 });
      wrapper.find('[data-action="Next"]').simulate('click');
      expect(wrapper.state('slideIndex')).toBe(0);
    });
  });
  it('renders the current slide as a CarouselSlide', () => {
    let slideProps;
    slideProps = wrapper.find(CarouselSlide).props();
    expect(slideProps).toEqual(slides[0]);
    wrapper.setState({ slideIndex: 1 });
    slideProps = wrapper.find(CarouselSlide).props();
    expect(slideProps).toEqual(slides[1]);
  });
});
