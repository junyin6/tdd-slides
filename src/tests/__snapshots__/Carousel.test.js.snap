// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackButton renders correctly 1`] = `
.c0 {
  position: absolute;
  margin: 0.5rem;
  padding: 1rem;
  height: 600px;
  left: 0;
  top: 0;
  border: none;
  background: rgba(0,0,0,0.3);
  font-size: 2rem;
  color: #f8f9fa;
}

<button
  className="c0"
>
  buttonText
</button>
`;

exports[`Carousel core Component renders correctly 1`] = `
<Carousel__CarouselWrapper
  defaultBackButton={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Carousel__BackButton-f9wgym-1",
        "isStatic": true,
        "rules": Array [
          "position:absolute;margin:0.5rem;padding:1rem;height:600px;left:0;top:0;border:none;background:rgba(0,0,0,0.3);font-size:2rem;color:#f8f9fa;",
        ],
      },
      "displayName": "Carousel__BackButton",
      "render": [Function],
      "styledComponentId": "Carousel__BackButton-f9wgym-1",
      "target": [Function],
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
      Symbol(Symbol.hasInstance): [Function],
    }
  }
  defaultCarouselWrapper={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Carousel__CarouselWrapper-f9wgym-0",
        "isStatic": true,
        "rules": Array [
          "position:relative;background-color:black;color:#f8f9fa;",
        ],
      },
      "displayName": "Carousel__CarouselWrapper",
      "render": [Function],
      "styledComponentId": "Carousel__CarouselWrapper-f9wgym-0",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  defaultForwardButton={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Carousel__ForwardButton-f9wgym-2",
        "isStatic": true,
        "rules": Array [
          "position:absolute;margin:0.5rem;padding:1rem;height:600px;right:0;top:0;border:none;background:rgba(0,0,0,0.3);font-size:2rem;color:#f8f9fa;",
        ],
      },
      "displayName": "Carousel__ForwardButton",
      "render": [Function],
      "styledComponentId": "Carousel__ForwardButton-f9wgym-2",
      "target": [Function],
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
      Symbol(Symbol.hasInstance): [Function],
    }
  }
>
  <CarouselSlide
    Img={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "CarouselSlide__Img-ds4ujd-1",
          "isStatic": false,
          "rules": Array [
            "object-fit:cover;width:100%;height:",
            [Function],
            ";",
          ],
        },
        "displayName": "CarouselSlide__Img",
        "render": [Function],
        "styledComponentId": "CarouselSlide__Img-ds4ujd-1",
        "target": "img",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    attribution="Uno Pizzeria"
    description="Slide 1"
    imgHeight={600}
    imgUrl="https://example.com/slide1.png"
  />
  <Carousel__BackButton
    data-action="Prev"
    onClick={[Function]}
  >
    &lt;
  </Carousel__BackButton>
  <Carousel__ForwardButton
    data-action="Next"
    onClick={[Function]}
  >
    &gt;
  </Carousel__ForwardButton>
</Carousel__CarouselWrapper>
`;

exports[`CarouselWrapper renders correctly 1`] = `
.c0 {
  position: relative;
  background-color: black;
  color: #f8f9fa;
}

<div
  className="c0"
/>
`;

exports[`ForwardButton renders correctly 1`] = `
.c0 {
  position: absolute;
  margin: 0.5rem;
  padding: 1rem;
  height: 600px;
  right: 0;
  top: 0;
  border: none;
  background: rgba(0,0,0,0.3);
  font-size: 2rem;
  color: #f8f9fa;
}

<button
  className="c0"
>
  buttonText
</button>
`;
